app/src/main/java/com/framgia/englishconversation/widget/ZoomableImageView.java

```diff
@@ -0,0 +1,207 @@
+                    case MotionEvent.ACTION_MOVE:
+                        if (mMode == ZOOM || (mMode == DRAG && mSaveScale > MIN_SCALE)) {
+                            float deltaX = curr.x - mLast.x;
+                            float deltaY = curr.y - mLast.y;
+                            if (Math.round(mOrigWidth * mSaveScale) < mWidth) {
+                                deltaX = 0;
+                                deltaY = setDeltaY(y, deltaY);
+                            } else if (Math.round(mOrigHeight * mSaveScale) < mHeight) {
+                                deltaY = 0;
+                                deltaX = setDeltaX(x, deltaX);
+                            } else {
+                                deltaX = setDeltaX(x, deltaX);
+                                deltaY = setDeltaY(y, deltaY);
+                            }
+                            mMatrix.postTranslate(deltaX, deltaY);
+                            mLast.set(curr.x, curr.y);
+                        }
+                        break;
+                    case MotionEvent.ACTION_UP:
+                    case MotionEvent.ACTION_POINTER_UP:
+                        mMode = NONE;
+                        break;
+                }
+                setImageMatrix(mMatrix);
+                invalidate();
+                return true;
+            }
+        });
+    }
+
+    @Override
+    public void setImageBitmap(Bitmap bm) {
+        super.setImageBitmap(bm);
+        mBmWidth = bm.getWidth();
+        mBmHeight = bm.getHeight();
+    }
+
+    /**
+     * ScaleListener
+     */
+
+    private class ScaleListener extends ScaleGestureDetector.SimpleOnScaleGestureListener {
+
+        @Override
+        public boolean onScaleBegin(ScaleGestureDetector detector) {
+            mMode = ZOOM;
+            return true;
+        }
+
+        @Override
+        public boolean onScale(ScaleGestureDetector detector) {
+            float mScaleFactor = detector.getScaleFactor();
+            float origScale = mSaveScale;
+            mSaveScale *= mScaleFactor;
+            if (mSaveScale > MAX_SCALE) {
+                mSaveScale = MAX_SCALE;
+                mScaleFactor = MAX_SCALE / origScale;
+            } else if (mSaveScale < MIN_SCALE) {
+                mSaveScale = MIN_SCALE;
+                mScaleFactor = MIN_SCALE / origScale;
+            }
+            mRight = mWidth * mSaveScale - mWidth - (2 * mRedundantXSpace * mSaveScale);
+            mBottom = mHeight * mSaveScale - mHeight - (2 * mRedundantYSpace * mSaveScale);
+            if (mOrigWidth * mSaveScale <= mWidth || mOrigHeight * mSaveScale <= mHeight) {
+                mMatrix.postScale(mScaleFactor, mScaleFactor, mWidth / 2, mHeight / 2);
+                mMatrix.getValues(mValues);
+                float x = mValues[Matrix.MTRANS_X];
+                float y = mValues[Matrix.MTRANS_Y];
+                if (mScaleFactor < 1) {
+                    if (mScaleFactor < 1) {
```

Comment:
WARNING: [CollapsibleIfStatements] These nested if statements could be combined
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/basic.html#CollapsibleIfStatements